google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/main/java/com/google/cloud/spanner/jdbc/ReadWriteTransaction.java

```diff
@@ -0,0 +1,763 @@
+  }
+
+  @Override
+  public boolean isReadOnly() {
+    return false;
+  }
+
+  @Override
+  void checkValidTransaction() {
+    ConnectionPreconditions.checkState(
+        state == UnitOfWorkState.STARTED,
+        "This transaction has status "
+        + state.name()
+        + ", only "
+        + UnitOfWorkState.STARTED
+        + " is allowed.");
+    ConnectionPreconditions.checkState(
+        !timedOutOrCancelled,
+        "The last statement of this transaction timed out or was cancelled. "
+            + "The transaction is no longer usable. "
+            + "Rollback the transaction and start a new one.");
+    if (txManager.getState() == null) {
+      transactionStarted = Timestamp.now();
+      txContext = txManager.begin();
+    }
+    if (txManager.getState()
+        != com.google.cloud.spanner.TransactionManager.TransactionState.STARTED) {
+      throw SpannerExceptionFactory.newSpannerException(
+          ErrorCode.FAILED_PRECONDITION,
+          String.format("Invalid transaction state: %s", txManager.getState()));
+    }
+  }
+
+  @Override
+  TransactionContext getReadContext() {
+    ConnectionPreconditions.checkState(txContext != null, "Missing transaction context");
+    return txContext;
+  }
+
+  @Override
+  public Timestamp getReadTimestamp() {
+    throw SpannerExceptionFactory.newSpannerException(
+        ErrorCode.FAILED_PRECONDITION,
+        "There is no read timestamp available for read/write transactions.");
+  }
+
+  @Override
+  public Timestamp getReadTimestampOrNull() {
+    return null;
+  }
+
+  private boolean hasCommitTimestamp() {
+    return txManager.getState()
+        == com.google.cloud.spanner.TransactionManager.TransactionState.COMMITTED;
+  }
+
+  @Override
+  public Timestamp getCommitTimestamp() {
+    ConnectionPreconditions.checkState(hasCommitTimestamp(), "This transaction has not committed.");
+    return txManager.getCommitTimestamp();
+  }
+
+  @Override
+  public Timestamp getCommitTimestampOrNull() {
+    return hasCommitTimestamp() ? txManager.getCommitTimestamp() : null;
+  }
+
+  @Override
+  public void executeDdl(ParsedStatement ddl) {
+    throw SpannerExceptionFactory.newSpannerException(
+        ErrorCode.FAILED_PRECONDITION,
+        "DDL-statements are not allowed inside a read/write transaction.");
+  }
+
+  private void handlePossibleInvalidatingException(SpannerException e) {
+    if (e.getErrorCode() == ErrorCode.DEADLINE_EXCEEDED
+        || e.getErrorCode() == ErrorCode.CANCELLED) {
+      this.timedOutOrCancelled = true;
+    }
+  }
+
+  @Override
+  public ResultSet executeQuery(
+      final ParsedStatement statement,
+      final AnalyzeMode analyzeMode,
+      final QueryOption... options) {
+    Preconditions.checkArgument(statement.isQuery(), "Statement is not a query");
+    checkValidTransaction();
+    try {
+      if (retryAbortsInternally) {
+        return asyncExecuteStatement(
+            statement,
+            new Callable<ResultSet>() {
```

Comment:
minor - can address in a future PR: let's extract out the Callable into a descriptive variable